src/LibraryManager/Providers/BaseProvider.cs

```diff
@@ -247,48 +247,92 @@ public async Task<OperationResult<LibraryInstallationGoalState>> GetInstallation
 
         private OperationResult<LibraryInstallationGoalState> GenerateGoalState(ILibraryInstallationState desiredState, ILibrary library)
         {
+            var mappings = new List<FileMapping>(desiredState.FileMappings ?? []);
             List<IError> errors = null;
-
-            if (string.IsNullOrEmpty(desiredState.DestinationPath))
-            {
-                return OperationResult<LibraryInstallationGoalState>.FromError(PredefinedErrors.DestinationNotSpecified(desiredState.Name));
-            }
-
-            IEnumerable<string> outFiles;
-            if (desiredState.Files == null || desiredState.Files.Count == 0)
+            if (desiredState.Files is { Count: > 0 })
             {
-                outFiles = library.Files.Keys;
+                mappings.Add(new FileMapping { Destination = desiredState.DestinationPath, Files = desiredState.Files });
             }
-            else
+            else if (desiredState.FileMappings is null or { Count: 0 })
             {
-                outFiles = FileGlobbingUtility.ExpandFileGlobs(desiredState.Files, library.Files.Keys);
+                // no files specified and no file mappings => include all files
+                mappings.Add(new FileMapping { Destination = desiredState.DestinationPath });
             }
 
             Dictionary<string, string> installFiles = new();
-            if (library.GetInvalidFiles(outFiles.ToList()) is IReadOnlyList<string> invalidFiles
-                && invalidFiles.Count > 0)
-            {
-                errors ??= [];
-                errors.Add(PredefinedErrors.InvalidFilesInLibrary(desiredState.Name, invalidFiles, library.Files.Keys));
-            }
 
-            foreach (string outFile in outFiles)
+            foreach (FileMapping fileMapping in mappings)
             {
-                // strip the source prefix
-                string destinationFile = Path.Combine(HostInteraction.WorkingDirectory, desiredState.DestinationPath, outFile);
-                if (!FileHelpers.IsUnderRootDirectory(destinationFile, HostInteraction.WorkingDirectory))
+                // if Root is not specified, assume it's the root of the library
+                string mappingRoot = fileMapping.Root ?? string.Empty;
+                // if Destination is not specified, inherit from the library entry
+                string destination = fileMapping.Destination ?? desiredState.DestinationPath;
+
+                if (destination is null)
+                {
+                    errors ??= [];
+                    string libraryId = LibraryNamingScheme.GetLibraryId(desiredState.Name, desiredState.Version);
+                    errors.Add(PredefinedErrors.DestinationNotSpecified(libraryId));
```

Comment:
Is there a case for v1 manifests where (if they followed the schema) they could encounter this error?

Reply:
The v1 schema was declared such that either a defaultDestination must be set, or every library must set a destination.  But it got too complicated to be oneOf (defaultDestination, or every library sets a destination,  or (every library without fileMappings sets a destination and every library with fileMappings has a fileMapping destination)).